airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -145,68 +173,76 @@ public synchronized void scanLibrary() {
+            Map<String, Artist> artists = new ConcurrentHashMap<>();
+            Map<MediaFile, Album> albums = new ConcurrentHashMap<>();
             Genres genres = new Genres();
 
-            scanCount = 0;
+            scanCount.set(0);
             statistics.reset();
 
             mediaFileService.setMemoryCacheEnabled(false);
             indexManager.startIndexing();
 
-            mediaFileService.clearMemoryCache();
-
             // Recurse through all files on disk.
-            for (MusicFolder musicFolder : settingsService.getAllMusicFolders()) {
-                MediaFile root = mediaFileService.getMediaFile(musicFolder.getPath(), false);
-                scanFile(root, musicFolder, lastScanned, albumCount, genres, false);
-            }
-
+            ForkJoinTask.invokeAll(
+                    settingsService.getAllMusicFolders()
+                        .parallelStream()
+                        .map(musicFolder -> ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(musicFolder.getPath(), false), musicFolder, lastScanned, albumCount, artists, albums, genres, false)))
+                        .collect(Collectors.toList()));
+            
             // Scan podcast folder.
             File podcastFolder = new File(settingsService.getPodcastFolder());
             if (podcastFolder.exists()) {
-                scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
-                         lastScanned, albumCount, genres, true);
+                ForkJoinTask.invokeAll(ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
+                         lastScanned, albumCount, artists, albums, genres, true)));
             }
 
             LOG.info("Scanned media library with " + scanCount + " entries.");
 
-            LOG.info("Marking non-present files.");
-            mediaFileDao.markNonPresent(lastScanned);
-            LOG.info("Marking non-present artists.");
-            artistDao.markNonPresent(lastScanned);
-            LOG.info("Marking non-present albums.");
-            albumDao.markNonPresent(lastScanned);
-
             // Update statistics
             statistics.incrementArtists(albumCount.size());
-            for (Integer albums : albumCount.values()) {
-                statistics.incrementAlbums(albums);
-            }
-
-            // Update genres
-            mediaFileDao.updateGenres(genres.getGenres());
+            statistics.incrementAlbums(albumCount.values().parallelStream().mapToInt(x -> x.get()).sum());
+            
+            ForkJoinTask.invokeAll(
+                    Streams.concat(
+                            albums.values().parallelStream().map(a -> ForkJoinTask.adapt(() -> albumDao.createOrUpdateAlbum(a))),
+                            artists.values().parallelStream().map(a -> ForkJoinTask.adapt(() -> artistDao.createOrUpdateArtist(a))),
```

Comment:
Persist artists and albums first (we built an in-memory model and need to commit at the end)